File: csharp/1186.maximum-subarray-sum-with-one-deletion.cs
/*
 * @lc app=leetcode id=1186 lang=csharp
 *
 * [1186] Maximum Subarray Sum with One Deletion
 */

using System;

// @lc code=start
public class Solution {
    public int MaximumSum(int[] arr) {
        int n = arr.Length, ret = int.MinValue;
        int[,] dp = new int[2, n + 1];
        dp[0, 0] = -100000;
        dp[1, 0] = -100000;
        for (int i = 0; i < n; ++i) {
            int curr = arr[i];
            dp[0, i + 1] = curr + Math.Max(0, dp[0, i]);
            dp[1, i + 1] = Math.Max(dp[0, i], curr + dp[1, i]);
            ret = Math.Max(ret, Math.Max(dp[0, i + 1], dp[1, i + 1]));
        }
        return ret;
    }
}
// @lc code=end

File: java/1186.maximum-subarray-sum-with-one-deletion.java
/*
 * @lc app=leetcode id=1186 lang=java
 *
 * [1186] Maximum Subarray Sum with One Deletion
 */

// @lc code=start
class Solution {
    public int maximumSum(int[] arr) {
        int ret = Integer.MIN_VALUE, del = -1000000, noDel = -1000000;
        for (int num : arr) {
            int nextNoDel = num + Math.max(0, noDel), nextDel = Math.max(noDel, num + del);
            ret = Math.max(ret, Math.max(nextNoDel, nextDel));
            del = nextDel;
            noDel = nextNoDel;
        }
        return ret;
    }
}
// @lc code=end

